Source/utils/stdcompat/optional.hpp

```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#ifdef _MSC_VER
```

Comment:
Perhaps we should update MSVC version on CI instead

`__cplusplus` is correctly defined as of Visual Studio 2017 version 15.7 Preview 3

https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/

Reply:
`__has_include` is also only supported since MSVC 2017 15.3.

Reply:
Good news: AppVeyor already has a recent enough version!
What we're missing is the `/Zc:__cplusplus` compilation flag, so let's add it instead

> You need to compile with the /Zc:__cplusplus switch to see the updated value of the __cplusplus macro. We tried updating the macro by default and discovered that a lot of code doesn’t compile correctly when we change the value of __cplusplus. We’ll continue to require use of the /Zc:__cplusplus switch for all minor versions of MSVC in the 19.xx family.

Reply:
Trying that.

Reply:
I'm fine with doing the switch. It's free and 2017 is already so recent that it's not gaining us any street cred.

Reply:
Nope: `Command line error D8021: invalid numeric argument '/W0 /Zc:__cplusplus'`.
Am I doing something wrong: `target_compile_options(${BIN_TARGET} PRIVATE "/W0 /Zc:__cplusplus")`?